### PR TITLE
[FlagGems Operator Development Competition]add conv_transpose2d operator

### PR DESCRIPTION
**Pull request by withdrawn919**
## PR Category
 Operator

## Type of Change
 New Feature

## Description
  - Add conv_transpose2d operator.
  - Add conv_transpose2d_backward operator.
  - Add accuracy tests for conv_transpose2d and conv_transpose2d_backward.
  - Add benchmark for conv_transpose2d and conv_transpose2d_backward.
  - Optimize conv_transpose2d and conv_transpose2d_backward performance for core benchmark shapes.

## Test
`pytest tests/test_conv_transpose2d.py -s`

```
========================================================================= test session starts =========================================================================
platform linux -- Python 3.12.13, pytest-9.0.3, pluggy-1.6.0
rootdir: /home/chouti/workspace/FlagGems
configfile: pytest.ini
collected 20 items                                                                                                                                                    

tests/test_conv_transpose2d.py ....................

========================================================================= 20 passed in 9.47s ==========================================================================

```
## Performance
`pytest benchmark/test_conv_transpose2d.py -v -s`

```
============================= test session starts ==============================
platform linux -- Python 3.12.13, pytest-9.0.3, pluggy-1.6.0 -- /home/chouti/miniconda3/envs/torch271/bin/python3.12
cachedir: .pytest_cache
rootdir: /home/chouti/workspace/FlagGems
configfile: pytest.ini
collecting ... collected 2 items

benchmark/test_conv_transpose2d.py::test_conv_transpose2d 
Operator: conv_transpose2d  Performance Test (dtype=torch.float16, mode=kernel,level=comprehensive)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               1.807040            0.224256               8.058          ([torch.Size([16, 32, 64, 64]), torch.Size([32, 32, 3, 3]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 1, 'dilation': 1})
SUCCESS               0.671552            0.139264               4.822          ([torch.Size([16, 64, 32, 32]), torch.Size([64, 32, 4, 4]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.558240            0.112640               4.956          ([torch.Size([8, 64, 64, 64]), torch.Size([64, 32, 3, 3]), torch.Size([32])], {'stride': 1, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.462848            0.055296               8.370          ([torch.Size([8, 64, 32, 32]), torch.Size([64, 16, 3, 3]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 2, 'dilation': 1})


Operator: conv_transpose2d  Performance Test (dtype=torch.float32, mode=kernel,level=comprehensive)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               2.213888            1.882112               1.176          ([torch.Size([16, 32, 64, 64]), torch.Size([32, 32, 3, 3]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 1, 'dilation': 1})
SUCCESS               0.838656            0.823296               1.019          ([torch.Size([16, 64, 32, 32]), torch.Size([64, 32, 4, 4]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.715776            0.558080               1.283          ([torch.Size([8, 64, 64, 64]), torch.Size([64, 32, 3, 3]), torch.Size([32])], {'stride': 1, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.494592            0.217088               2.278          ([torch.Size([8, 64, 32, 32]), torch.Size([64, 16, 3, 3]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 2, 'dilation': 1})


Operator: conv_transpose2d  Performance Test (dtype=torch.bfloat16, mode=kernel,level=comprehensive)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               1.994752            0.249856               7.984          ([torch.Size([16, 32, 64, 64]), torch.Size([32, 32, 3, 3]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 1, 'dilation': 1})
SUCCESS               0.836608            0.138240               6.052          ([torch.Size([16, 64, 32, 32]), torch.Size([64, 32, 4, 4]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.588064            0.119808               4.908          ([torch.Size([8, 64, 64, 64]), torch.Size([64, 32, 3, 3]), torch.Size([32])], {'stride': 1, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.466944            0.057056               8.184          ([torch.Size([8, 64, 32, 32]), torch.Size([64, 16, 3, 3]), torch.Size([32])], {'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 2, 'dilation': 1})

PASSED
benchmark/test_conv_transpose2d.py::test_conv_transpose2d_backward 
Operator: conv_transpose2d_backward  Performance Test (dtype=torch.float16, mode=kernel,level=comprehensive)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               0.547840            0.160768               3.408          ([torch.Size([16, 32, 64, 64])], {'weight': torch.Size([32, 32, 3, 3]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 1, 'dilation': 1})
SUCCESS               0.349184            0.119904               2.912          ([torch.Size([16, 64, 32, 32])], {'weight': torch.Size([64, 32, 4, 4]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.255904            0.094208               2.716          ([torch.Size([8, 64, 64, 64])], {'weight': torch.Size([64, 32, 3, 3]), 'bias': torch.Size([32]), 'stride': 1, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.242688            0.031808               7.630          ([torch.Size([8, 64, 32, 32])], {'weight': torch.Size([64, 16, 3, 3]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 2, 'dilation': 1})


Operator: conv_transpose2d_backward  Performance Test (dtype=torch.float32, mode=kernel,level=comprehensive)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               0.977920            0.758784               1.289          ([torch.Size([16, 32, 64, 64])], {'weight': torch.Size([32, 32, 3, 3]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 1, 'dilation': 1})
SUCCESS               0.525312            0.276480               1.900          ([torch.Size([16, 64, 32, 32])], {'weight': torch.Size([64, 32, 4, 4]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.396288            0.303104               1.307          ([torch.Size([8, 64, 64, 64])], {'weight': torch.Size([64, 32, 3, 3]), 'bias': torch.Size([32]), 'stride': 1, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.299008            0.107520               2.781          ([torch.Size([8, 64, 32, 32])], {'weight': torch.Size([64, 16, 3, 3]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 2, 'dilation': 1})


Operator: conv_transpose2d_backward  Performance Test (dtype=torch.bfloat16, mode=kernel,level=comprehensive)
Status       Torch Latency (ms)    Gems Latency (ms)         Gems Speedup          Size Detail
-----------------------------------------------------------------------------------------------
SUCCESS               0.625664            0.162816               3.843          ([torch.Size([16, 32, 64, 64])], {'weight': torch.Size([32, 32, 3, 3]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 1, 'dilation': 1})
SUCCESS               0.335872            0.124928               2.689          ([torch.Size([16, 64, 32, 32])], {'weight': torch.Size([64, 32, 4, 4]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.276480            0.094208               2.935          ([torch.Size([8, 64, 64, 64])], {'weight': torch.Size([64, 32, 3, 3]), 'bias': torch.Size([32]), 'stride': 1, 'padding': 1, 'output_padding': 0, 'groups': 1, 'dilation': 1})
SUCCESS               0.262144            0.032672               8.024          ([torch.Size([8, 64, 32, 32])], {'weight': torch.Size([64, 16, 3, 3]), 'bias': torch.Size([32]), 'stride': 2, 'padding': 1, 'output_padding': 1, 'groups': 2, 'dilation': 1})

PASSED

============================== 2 passed in 38.44s ==============================

```

### CODE REVIEW

File: benchmark/test_conv_transpose2d.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    cuda_matmul_allow_tf32 = torch.backends.cuda.matmul.allow_tf32
+    torch.backends.cudnn.enabled = False
+    torch.backends.cudnn.allow_tf32 = False
+    torch.backends.cuda.matmul.allow_tf32 = True
+    try:
+        yield
+    finally:
+        torch.backends.cudnn.enabled = cudnn_enabled
+        torch.backends.cudnn.allow_tf32 = cudnn_allow_tf32
+        torch.backends.cuda.matmul.allow_tf32 = cuda_matmul_allow_tf32
```
Comment (tengqm) [inline benchmark/test_conv_transpose2d.py]: I'm unsure that these hacks work across different torch versions, and I'm worrying that these tunings won't work on non-NVIDIA chips.

Comment (withdrawn919) [reply benchmark/test_conv_transpose2d.py]: Restrict CUDA/cuDNN backend flag tuning in the conv_transpose2d benchmark to
  the NVIDIA backend only. Non-NVIDIA CUDA-compatible backends should not touch
  these flags.

  Use IEEE dot precision for conv_transpose2d on non-NVIDIA backends instead of
  deriving TF32 behavior from torch.backends.cuda.matmul.allow_tf32. Also reduce
  the fp32 stride-2 block size on non-NVIDIA backends to avoid shared-memory
  overuse on MetaX C550.

File: benchmark/test_conv_transpose2d.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+import pytest
+import torch
+
+import flag_gems
+from tests import accuracy_utils as utils
+
+from . import base
+
+
+@contextlib.contextmanager
```
Comment (tengqm) [inline benchmark/test_conv_transpose2d.py]: This `contextmanager` is an over-engineering in my opinion.
Can we revise this into a simple, boring implementation for the sake of readability and maintainability?

Comment (withdrawn919) [reply benchmark/test_conv_transpose2d.py]: Replace the contextmanager-based benchmark backend flag setup with explicit save/restore helpers and try/finally blocks. This keeps the benchmark behavior unchanged while improving readability and maintainability.

**Comment by douxetpur** [comment]
Hi, thank you for your participation in the FlagOS Open Computing Competition 🙏

After review, we have chosen to move forward with a different implementation for this operator. As a result, this PR will be closed.

We appreciate your time and contribution, and hope to see more of your submissions in the future.
